File: wf-api-integration/references/account-inquiry/notify-vostro/golang/client/notify_vostro_handler_test.go
package client

import (
	"fmt"
	"testing"

	"{moduleName}/wf/model/request"
)

// ==================== NotifyVostro Handler Tests ====================

// TestHandleNotification_FundingSuccess tests handling a funding success notification
func TestHandleNotification_FundingSuccess(t *testing.T) {
	handler := NewNotifyVostroHandler()

	// Track if callback was invoked
	callbackInvoked := false
	handler.OnFundingSuccess = func(req *request.NotifyVostroRequest) {
		callbackInvoked = true
		fmt.Printf("  OnFundingSuccess callback: fundingId=%s, amount=%d %s, beneficiary=%s\n",
			req.FundingID,
			req.BalanceChangeAmount.Value, req.BalanceChangeAmount.Currency,
			req.BeneficiaryAccount.BeneficiaryBankAccountNo)
	}

	body := []byte(`{
		"fundingId": "20260421000000001",
		"balanceResult": {
			"resultCode": "SUCCESS",
			"resultStatus": "S",
			"resultMessage": "Success"
		},
		"payerBankAccount": {
			"payerBankAccountNo": "622848*****1234",
			"payerBankName": "Bank of China"
		},
		"beneficiaryAccount": {
			"beneficiaryBankAccountNo": "VA0000001234567890"
		},
		"balanceChangeAmount": {
			"currency": "USD",
			"value": 10000
		},
		"balanceChangeTime": "2026-04-21T12:01:01+08:00",
		"remitInfo": "Test funding"
	}`)

	fmt.Println("====== TestHandleNotification_FundingSuccess ======")

	resp := handler.HandleNotification(body)

	fmt.Printf("  Response: resultStatus=%s, resultCode=%s\n",
		resp.Result.ResultStatus, resp.Result.ResultCode)

	if resp.Result.ResultStatus != "S" {
		t.Errorf("expected resultStatus=S, got %s", resp.Result.ResultStatus)
	}
	if resp.Result.ResultCode != "SUCCESS" {
		t.Errorf("expected resultCode=SUCCESS, got %s", resp.Result.ResultCode)
	}
	if !callbackInvoked {
		t.Error("expected OnFundingSuccess callback to be invoked")
	}

	fmt.Println("====================================================")
}

// TestHandleNotification_RefundSuccess tests handling a refund success notification
func TestHandleNotification_RefundSuccess(t *testing.T) {
	handler := NewNotifyVostroHandler()

	callbackInvoked := false
	handler.OnRefundSuccess = func(req *request.NotifyVostroRequest) {
		callbackInvoked = true
		fmt.Printf("  OnRefundSuccess callback: fundingId=%s, amount=%d %s\n",
			req.FundingID,
			req.BalanceChangeAmount.Value, req.BalanceChangeAmount.Currency)
	}

	body := []byte(`{
		"fundingId": "20260421000000002",
		"balanceResult": {
			"resultCode": "REFUND",
			"resultStatus": "S",
			"resultMessage": "Refund Success"
		},
		"payerBankAccount": {
			"payerBankAccountNo": "622848*****5678",
			"payerBankName": "ICBC"
		},
		"beneficiaryAccount": {
			"beneficiaryBankAccountNo": "VA0000009876543210"
		},
		"balanceChangeAmount": {
			"currency": "EUR",
			"value": 5000
		},
		"balanceChangeTime": "2026-04-21T14:30:00+08:00"
	}`)

	fmt.Println("====== TestHandleNotification_RefundSuccess ======")

	resp := handler.HandleNotification(body)

	fmt.Printf("  Response: resultStatus=%s, resultCode=%s\n",
		resp.Result.ResultStatus, resp.Result.ResultCode)

	if resp.Result.ResultStatus != "S" {
		t.Errorf("expected resultStatus=S, got %s", resp.Result.ResultStatus)
	}
	if !callbackInvoked {
		t.Error("expected OnRefundSuccess callback to be invoked")
	}

	fmt.Println("===================================================")
}

File: wf-api-integration/references/account-inquiry/notify-vostro/java/client/NotifyVostroHandlerTest.java
/**
 * Alipay.com Inc.
 * Copyright (c) 2004-2026 All Rights Reserved.
 */
package {basePackage}.wf;

import {basePackage}.wf.client.account.NotifyVostroHandler;
import {basePackage}.wf.model.response.NotifyVostroResponse;
import org.junit.Before;
import org.junit.Test;

/**
 * NotifyVostroHandler 单元测试。
 *
 * <p>模拟万里汇发送的 notifyVostro 回调通知，验证处理器的解析和响应逻辑。
 *
 * @author Qoder
 * @version NotifyVostroHandlerTest.java, v 0.1 2026-04-21
 */
public class NotifyVostroHandlerTest {

    private NotifyVostroHandler handler;

    @Before
    public void setUp() {
        handler = new NotifyVostroHandler();
    }

    // ==================== Funding Success Tests ====================

    /**
     * 测试处理垫付成功通知。
     */
    @Test
    public void testHandleFundingSuccess() {
        String requestBody = "{"
            + "\"fundingId\":\"20260421000000001\","
            + "\"balanceResult\":{"
            + "  \"resultCode\":\"SUCCESS\","
            + "  \"resultStatus\":\"S\","
            + "  \"resultMessage\":\"Success\""
            + "},"
            + "\"payerBankAccount\":{"
            + "  \"payerBankAccountNo\":\"622848*****1234\","
            + "  \"payerBankName\":\"Bank of China\""
            + "},"
            + "\"beneficiaryAccount\":{"
            + "  \"beneficiaryBankAccountNo\":\"VA0000001234567890\""
            + "},"
            + "\"balanceChangeAmount\":{"
            + "  \"currency\":\"USD\","
            + "  \"value\":10000"
            + "},"
            + "\"balanceChangeTime\":\"2026-04-21T12:01:01+08:00\","
            + "\"remitInfo\":\"Test funding\""
            + "}";

        System.out.println("====== testHandleFundingSuccess ======");
        System.out.println("Request body: " + requestBody);

        NotifyVostroResponse response = handler.handleNotification(requestBody);

        System.out.println("Response: " + response);
        System.out.println("ResultStatus: " + response.getResult().getResultStatus());
        System.out.println("ResultCode: " + response.getResult().getResultCode());

        assert "S".equals(response.getResult().getResultStatus());
        assert "SUCCESS".equals(response.getResult().getResultCode());

        System.out.println("======================================");
    }

    /**
     * 测试处理退款成功通知。
     */
    @Test
    public void testHandleRefundSuccess() {
        String requestBody = "{"
            + "\"fundingId\":\"20260421000000002\","
            + "\"balanceResult\":{"
            + "  \"resultCode\":\"REFUND\","
            + "  \"resultStatus\":\"S\","
            + "  \"resultMessage\":\"Refund Success\""
            + "},"
            + "\"payerBankAccount\":{"
            + "  \"payerBankAccountNo\":\"622848*****5678\","
            + "  \"payerBankName\":\"ICBC\""
            + "},"
            + "\"beneficiaryAccount\":{"
            + "  \"beneficiaryBankAccountNo\":\"VA0000009876543210\""
            + "},"
            + "\"balanceChangeAmount\":{"
            + "  \"currency\":\"EUR\","
            + "  \"value\":5000"
            + "},"
            + "\"balanceChangeTime\":\"2026-04-21T14:30:00+08:00\""
            + "}";

        System.out.println("====== testHandleRefundSuccess ======");
        System.out.println("Request body: " + requestBody);

        NotifyVostroResponse response = handler.handleNotification(requestBody);

        System.out.println("Response: " + response);
        System.out.println("ResultStatus: " + response.getResult().getResultStatus());
        System.out.println("ResultCode: " + response.getResult().getResultCode());

        assert "S".equals(response.getResult().getResultStatus());
        assert "SUCCESS".equals(response.getResult().getResultCode());

        System.out.println("=====================================");
    }
}
